vowpalwabbit/core/src/reductions/sender.cc

```diff
@@ -4,138 +4,152 @@
+{
+  for (auto& f : all.final_prediction_sink) { all.print_by_ref(f.get(), prediction, 0, info.tag, logger); }
+}
+
+void print_update_sender(VW::workspace& all, VW::shared_data& sd, const sent_example_info& info, float prediction)
+{
+  const bool should_print_driver_update = sd.weighted_examples() >= sd.dump_interval && !all.quiet;
+
+  if (should_print_driver_update)
   {
-    if (ns == VW::details::CONSTANT_NAMESPACE) { continue; }
-    VW::parsers::cache::details::cache_index(*b, ns);
-    VW::parsers::cache::details::cache_features(*b, ec.feature_space[ns], mask);
+    sd.print_update(*all.trace_message, all.holdout_set_off, all.current_pass, info.label.label, prediction,
+        info.num_features, all.progress_add, all.progress_arg);
   }
-  b->flush();
 }
 
 void receive_result(sender& s)
 {
-  float res;
-  float weight;
+  float prediction{};
+  float weight{};
 
-  get_prediction(s.socket_reader.get(), res, weight);
-  VW::example& ec = *s.delay_ring[s.received_index++ % s.all->example_parser->example_queue_limit];
-  ec.pred.scalar = res;
+  VW::details::get_prediction(s.socket_reader.get(), prediction, weight);
+  const auto& sent_info = s.delay_ring[s.received_index++ % s.all->example_parser->example_queue_limit];
 
-  auto& ld = ec.l.simple;
-  ec.loss = s.all->loss->get_loss(s.all->sd, ec.pred.scalar, ld.label) * ec.weight;
+  const auto& ld = sent_info.label;
+  const auto loss = s.all->loss->get_loss(s.all->sd, prediction, ld.label) * sent_info.weight;
 
-  VW::details::return_simple_example(*(s.all), nullptr, ec);
+  update_stats_sender(*(s.all->sd), sent_info, loss);
+  output_example_prediction_sender(*s.all, sent_info, prediction, s.all->logger);
+  print_update_sender(*s.all, *(s.all->sd), sent_info, prediction);
 }
 
-void learn(sender& s, VW::LEARNER::base_learner& /*unused*/, VW::example& ec)
+void send_example(sender& s, VW::LEARNER::base_learner& /* non_existent_base */, VW::example& ec)
 {
   if (s.received_index + s.all->example_parser->example_queue_limit / 2 - 1 == s.sent_index) { receive_result(s); }
 
   s.all->set_minmax(s.all->sd, ec.l.simple.label);
-  s.all->example_parser->lbl_parser.cache_label(
-      ec.l, ec.ex_reduction_features, *s.buf, "", false);  // send label information.
-  VW::parsers::cache::details::cache_tag(*s.buf, ec.tag);
-  send_features(s.buf, ec, static_cast<uint32_t>(s.all->parse_mask));
-  s.delay_ring[s.sent_index++ % s.all->example_parser->example_queue_limit] = &ec;
+  VW::parsers::cache::write_example_to_cache(
+      s.socket_output_buffer, &ec, s.all->example_parser->lbl_parser, s.all->parse_mask, s.cache_buffer);
+  s.socket_output_buffer.flush();
+  s.delay_ring[s.sent_index++ % s.all->example_parser->example_queue_limit] =
+      sent_example_info{ec.l.simple, ec.weight, ec.test_only, ec.get_num_features(), ec.tag};
 }
 
-void finish_example(VW::workspace& /*unused*/, sender& /*unused*/, VW::example& /*unused*/) {}
```

Comment:
so by removing this one, the default finish will be called?

Reply:
See line 149: 
https://github.com/VowpalWabbit/vowpal_wabbit/pull/4321/files/d8cfb12bdbab442f19670299e19a7c77a79d3257#diff-7b31f6c23efcb70f9ee3cf9cf62a32a08cd78dfa4ddfe11b79369b0080784560R149